#**Exploratory Data Cleaning & Preprocessing**
Author: Rishab Tiwari

Date: January 2026

Tools: Python, Pandas, NumPy

### **Project Objective**
The goal of this project is to transform a "raw" dataset into a "tidy" format suitable for analysis. This involves identifying and **fixing structural errors, duplicate data, incorrect spelling, outliers, incorrect data and data types, inconsistent formatting ,handling missing values,** and ensuring data consistency across all features.

#### **Steps of Data cleaning**
1.   Data Collaction and Inspaction.

  *   Check the meta data of the data
  *   Check the type of data in each column (Numerical, categorical, Date&Time, Text)
  * Missing Vlaues and Duplicate values
  * Data types
  * Inconsistencies


#### Techniques  

*   Summary statistics
*   Distribution plot
* .Info (Data types, Null and Not Null count)
* Frequency (Categorical data)
* Visual Inspection

In [ ]:
# Importing Essential Libraries for the data and cleaning process.
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# Replace 'path/to/your/file.csv' with the actual path to your CSV file
file_path = '/content/drive/My Drive/Practice_data/dirty_cafe_sales.csv'
df = pd.read_csv(file_path)
df.head(10)

Transaction ID      Item Quantity Price Per Unit Total Spent  \
0    TXN_1961373    Coffee        2              2           4   
1    TXN_4977031      Cake        4              3          12   
2    TXN_4271903    Cookie        4              1       ERROR   
3    TXN_7034554     Salad        2              5          10   
4    TXN_3160411    Coffee        2              2           4   
5    TXN_2602893  Smoothie        5              4          20   
6    TXN_4433211   UNKNOWN        3              3           9   
7    TXN_6699534  Sandwich        4              4          16   
8    TXN_4717867       NaN        5              3          15   
9    TXN_2064365  Sandwich        5              4          20   

   Payment Method  Location Transaction Date  
0     Credit Card  Takeaway       2023-09-08  
1            Cash  In-store       2023-05-16  
2     Credit Card  In-store       2023-07-19  
3         UNKNOWN   UNKNOWN       2023-04-27  
4  Digital Wallet  In-store       2023-06-11  
5     Credit Card       NaN       2023-03-31  
6           ERROR  Takeaway       2023-10-06  
7            Cash   UNKNOWN       2023-10-28  
8             NaN  Takeaway       2023-07-28  
9             NaN  In-store       2023-12-31

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Transaction ID    10000 non-null  object
 1   Item              9667 non-null   object
 2   Quantity          9862 non-null   object
 3   Price Per Unit    9821 non-null   object
 4   Total Spent       9827 non-null   object
 5   Payment Method    7421 non-null   object
 6   Location          6735 non-null   object
 7   Transaction Date  9841 non-null   object
dtypes: object(8)
memory usage: 625.1+ KB


## **Observation**
#### The data has 8 columns, named {Transaction ID, Item, Quantity, Price Per unit, Total Spent, Payment Methods, Location, Transaction date.} and 10000 rows.
* All the data is in object data type, it means Quantity, Price Per unit, Total Spent and transaction data has wrong data type.
*  Payment Method (7421 non-null) and Location (6735 non-null) column has most of the null values in the data.


In [ ]:
df.describe()

Transaction ID   Item Quantity Price Per Unit Total Spent  \
count           10000   9667     9862           9821        9827   
unique          10000     10        7              8          19   
top       TXN_9226047  Juice        5              3           6   
freq                1   1171     2013           2429         979   

        Payment Method  Location Transaction Date  
count             7421      6735             9841  
unique               5         4              367  
top     Digital Wallet  Takeaway          UNKNOWN  
freq              2291      3022              159

Because of the data type is object, we are unable to see the data summary using describe() function of pandas. we can not understand the true summary before converting the data into thier belonging data type.

In [ ]:
df.nunique()

Transaction ID      10000
Item                   10
Quantity                7
Price Per Unit          8
Total Spent            19
Payment Method          5
Location                4
Transaction Date      367
dtype: int64

### **Observation**
* Transaction ID has 10000 unique entries, which seems completly normal beacause we have 10000 rows of and every transaction is recoreded.
* Also, all the columns have believable amount of unique entries.

**To understand there uniquenes we will drill down furthur to confirm if they truely contain unique values of it is also counting null values.**

In [ ]:
unique_transaction = df['Transaction ID'].unique()
print(unique_transaction)

['TXN_1961373' 'TXN_4977031' 'TXN_4271903' ... 'TXN_5255387' 'TXN_7695629'
 'TXN_6170729']


In [ ]:
unique_itemes = df['Item'].unique()
print(unique_itemes)

['Coffee' 'Cake' 'Cookie' 'Salad' 'Smoothie' 'UNKNOWN' 'Sandwich' nan
 'ERROR' 'Juice' 'Tea']


In [ ]:
unique_quantities = df['Quantity'].unique()
print(unique_quantities)

['2' '4' '5' '3' '1' 'ERROR' 'UNKNOWN' nan]


In [ ]:
unique_price_per_unit = df['Price Per Unit'].unique()
print(unique_price_per_unit)

['2' '3' '1' '5' '4' '1.5' nan 'ERROR' 'UNKNOWN']


In [ ]:
unique_total_spent = df['Total Spent'].unique()
print(unique_total_spent)

['4' '12' 'ERROR' '10' '20' '9' '16' '15' '25' '8' '5' '3' '6' nan
 'UNKNOWN' '2' '1' '7.5' '4.5' '1.5']


In [ ]:
unique_Payment_Method = df['Payment Method'].unique()
print(unique_Payment_Method)

['Credit Card' 'Cash' 'UNKNOWN' 'Digital Wallet' 'ERROR' nan]


In [ ]:
unique_Location = df['Location'].unique()
print(unique_Location)

['Takeaway' 'In-store' 'UNKNOWN' nan 'ERROR']


In [ ]:

unique_Transaction_Date= df['Transaction Date'].unique()
print(unique_Transaction_Date)

['2023-09-08' '2023-05-16' '2023-07-19' '2023-04-27' '2023-06-11'
 '2023-03-31' '2023-10-06' '2023-10-28' '2023-07-28' '2023-12-31'
 '2023-11-07' 'ERROR' '2023-05-03' '2023-06-01' '2023-03-21' '2023-11-15'
 '2023-06-10' '2023-02-24' '2023-03-25' '2023-01-15' '2023-04-04'
 '2023-03-30' '2023-12-01' '2023-09-18' '2023-06-03' '2023-12-13'
 '2023-04-20' '2023-04-10' '2023-03-11' '2023-06-02' '2023-11-06'
 '2023-08-15' '2023-10-09' '2023-05-28' '2023-07-17' '2023-04-29'
 '2023-06-08' '2023-06-29' '2023-04-17' '2023-12-22' '2023-01-10'
 '2023-10-02' '2023-02-23' '2023-03-22' '2023-11-03' '2023-03-02'
 '2023-06-26' '2023-05-02' '2023-09-05' '2023-01-08' '2023-03-15'
 '2023-11-25' '2023-12-05' '2023-03-19' '2023-06-27' '2023-04-19'
 '2023-10-07' '2023-09-30' '2023-05-27' '2023-11-18' '2023-10-20'
 '2023-10-03' '2023-10-27' '2023-04-06' '2023-01-31' '2023-12-08'
 '2023-06-19' '2023-12-14' '2023-07-16' '2023-02-22' nan '2023-06-15'
 '2023-12-09' '2023-04-18' '2023-10-29' '2023-04-30' '2023-04-02

## **Observation**

* **Item** column should resemble the item name which is bought by the customers like ['Coffee' 'Cake' 'Cookie' 'Salad' 'Smoothie' 'Sandwich''Juice' 'Tea'], but also contains (UNKNOWN, nan, ERROR).

* **Quantity** column has some unique values that resembels null and error values of the data. it should contain data that resembels number of items bought, but it also contains ('ERROR' 'UNKNOWN' nan). that is also the reason its data type is object.

* **Price Per Unit** should contain price per unit like ['2' '3' '1' '5' '4' '1.5'] but still it contains [ nan 'ERROR' 'UNKNOWN'].

* **Total Spent** should contain numerical values which is {Quantity x Price per unite} but also contains ['ERROR' nan
 'UNKNOWN'].

 * **Payment Method**  should contain object data type and information of using which method payment was done. example ['Credit Card' 'Cash' 'Digital Wallet'], but it also contains ['UNKNOWN' 'ERROR' nan].

*   **Location** contains the data what type of ordere was like ['Takeaway' 'In-store'] but it also has ['UNKNOWN' nan 'ERROR'].

*   **Transaction date** also has ERROR nan and UNKOWN in it.


We got to know that our data has some irrelevent values, Generally it it dropped or taken out to make data null value and error free.
But in this case i think we can fix data in some columns.

we will find a way to fill these values as much as possible.



In [ ]:
df.head(10)

Transaction ID      Item  Quantity  Price Per Unit  Total Spent  \
0    TXN_1961373    Coffee       2.0             2.0          4.0   
1    TXN_4977031      Cake       4.0             3.0         12.0   
2    TXN_4271903    Cookie       4.0             1.0          NaN   
3    TXN_7034554     Salad       2.0             5.0         10.0   
4    TXN_3160411    Coffee       2.0             2.0          4.0   
5    TXN_2602893  Smoothie       5.0             4.0         20.0   
6    TXN_4433211     Juice       3.0             3.0          9.0   
7    TXN_6699534  Sandwich       4.0             4.0         16.0   
8    TXN_4717867     Juice       5.0             3.0         15.0   
9    TXN_2064365  Sandwich       5.0             4.0         20.0   

   Payment Method  Location Transaction Date  
0     Credit Card  Takeaway       2023-09-08  
1            Cash  In-store       2023-05-16  
2     Credit Card  In-store       2023-07-19  
3         UNKNOWN   UNKNOWN       2023-04-27  
4  Digital Wallet  In-store       2023-06-11  
5     Credit Card       NaN       2023-03-31  
6           ERROR  Takeaway       2023-10-06  
7            Cash   UNKNOWN       2023-10-28  
8             NaN  Takeaway       2023-07-28  
9             NaN  In-store       2023-12-31

### **As we can see column "Item" contains names of food items and every food item has its price mentioned next to the quantity.**

- Item = [Coffee' 'Cake' 'Cookie' 'Salad' 'Smoothie' 'Sandwich'  'Juice' 'Tea']
- Price = [2,3,1,5,4,4,3,1.5]

Now, we have have Item and thier price, we can fill all the values where one of the element is missing.


- ### **We have a table with columns 'Item' and 'Price Per Unit', Some values are missing (NaN, UNKNOWN, ERROR)**

### **To Solve this problem**

#### **Logic Rules**
- Rule 1: Item is NaN/Unknown, Price Per Unit has a value
    - Action: Fill the Item based on the Price Per Unit
    - Example: If Item = NaN and Price Per Unit = 2, fill Item with "Coffee"



- Rule 2: Price Per Unit is NaN/Unknown, Item has a value

  - Action: Fill the Price Per Unit based on the Item
  - Example: If Item = "Salad" and Price Per Unit = NaN, fill Price Per Unit with 5

- Rule 3: Both Item and Price Per Unit are NaN/Unknown

  - Action: Leave both unchanged (no modifications)
  - Example: If Item = NaN and Price Per Unit = NaN, both remain NaN

In [ ]:
import pandas as pd
import numpy as np

def fill_missing_values(df):
    """
    Fill NaN, error, and unknown values in 'Item' and 'Price Per Unit' columns
    based on a predefined dictionary mapping.

    Parameters:
    df (pd.DataFrame): DataFrame with columns including 'Item' and 'Price Per Unit'

    Returns:
    pd.DataFrame: DataFrame with filled values
    """

    # Define the dictionary mapping
    items = ['Coffee', 'Cake', 'Cookie', 'Salad', 'Smoothie', 'Sandwich', 'Juice', 'Tea']
    prices = [2, 3, 1, 5, 4, 4, 3, 1.5]

    # Create bidirectional mappings
    item_to_price = dict(zip(items, prices))
    price_to_items = {}
    for item, price in item_to_price.items():
        if price not in price_to_items:
            price_to_items[price] = []
        price_to_items[price].append(item)

    # Create a copy to avoid modifying the original
    df_filled = df.copy()

    # Define values to consider as missing/invalid
    invalid_values = ['NaN', 'nan', 'UNKNOWN', 'ERROR', None, np.nan, 'unknown', 'error']

    def is_invalid(value):
        """Check if a value is invalid (nan, error, unknown, etc.)"""
        if pd.isna(value):
            return True
        if isinstance(value, str):
            return value.strip() in ['NaN', 'nan', 'UNKNOWN', 'ERROR', 'unknown', 'error']
        return False

    # Iterate through each row
    for idx in df_filled.index:
        item_value = df_filled.loc[idx, 'Item']
        price_value = df_filled.loc[idx, 'Price Per Unit']

        item_is_invalid = is_invalid(item_value)
        price_is_invalid = is_invalid(price_value)

        # Case 1: Both are invalid - leave unchanged
        if item_is_invalid and price_is_invalid:
            continue

        # Case 2: Item is invalid but Price Per Unit is valid
        elif item_is_invalid and not price_is_invalid:
            try:
                price_float = float(price_value)
                # Look up items matching this price
                if price_float in price_to_items:
                    matching_items = price_to_items[price_float]
                    # If multiple items have the same price, take the first one
                    # (or you could implement different logic)
                    df_filled.loc[idx, 'Item'] = matching_items[0]
            except (ValueError, TypeError):
                # Price value cannot be converted to float, skip
                pass

        # Case 3: Price Per Unit is invalid but Item is valid
        elif not item_is_invalid and price_is_invalid:
            # Check if the item is in our dictionary
            if item_value in item_to_price:
                df_filled.loc[idx, 'Price Per Unit'] = item_to_price[item_value]

    return df_filled



In [ ]:
df_filled = fill_missing_values(df)

# Display the results
print("Original DataFrame:")
print(df[['Transaction ID', 'Item', 'Price Per Unit']])
print("\nFilled DataFrame:")
print(df_filled[['Transaction ID', 'Item', 'Price Per Unit']])

# Or simply view df_filled
df_filled


Original DataFrame:
     Transaction ID      Item Price Per Unit
0       TXN_1961373    Coffee              2
1       TXN_4977031      Cake              3
2       TXN_4271903    Cookie              1
3       TXN_7034554     Salad              5
4       TXN_3160411    Coffee              2
...             ...       ...            ...
9995    TXN_7672686    Coffee              2
9996    TXN_9659401       NaN            NaN
9997    TXN_5255387    Coffee              2
9998    TXN_7695629    Cookie            NaN
9999    TXN_6170729  Sandwich              4

[10000 rows x 3 columns]

Filled DataFrame:
     Transaction ID      Item Price Per Unit
0       TXN_1961373    Coffee              2
1       TXN_4977031      Cake              3
2       TXN_4271903    Cookie              1
3       TXN_7034554     Salad              5
4       TXN_3160411    Coffee              2
...             ...       ...            ...
9995    TXN_7672686    Coffee              2
9996    TXN_9659401       NaN      

Transaction ID      Item Quantity Price Per Unit Total Spent  \
0       TXN_1961373    Coffee        2              2           4   
1       TXN_4977031      Cake        4              3          12   
2       TXN_4271903    Cookie        4              1       ERROR   
3       TXN_7034554     Salad        2              5          10   
4       TXN_3160411    Coffee        2              2           4   
...             ...       ...      ...            ...         ...   
9995    TXN_7672686    Coffee        2              2           4   
9996    TXN_9659401       NaN        3            NaN           3   
9997    TXN_5255387    Coffee        4              2           8   
9998    TXN_7695629    Cookie        3              1           3   
9999    TXN_6170729  Sandwich        3              4          12   

      Payment Method  Location Transaction Date  
0        Credit Card  Takeaway       2023-09-08  
1               Cash  In-store       2023-05-16  
2        Credit Card  In-store       2023-07-19  
3            UNKNOWN   UNKNOWN       2023-04-27  
4     Digital Wallet  In-store       2023-06-11  
...              ...       ...              ...  
9995             NaN   UNKNOWN       2023-08-30  
9996  Digital Wallet       NaN       2023-06-02  
9997  Digital Wallet       NaN       2023-03-02  
9998  Digital Wallet       NaN       2023-12-02  
9999            Cash  In-store       2023-11-07  

[10000 rows x 8 columns]

In [ ]:
import pandas as pd
import numpy as np

# 1. Create clean numeric Series (Internal variables, not changing df yet)
# errors='coerce' turns "UNKNOWN"/"ERROR" into NaN
qty_clean = pd.to_numeric(df_filled['Quantity'], errors='coerce')
price_clean = pd.to_numeric(df_filled['Price Per Unit'], errors='coerce')

# 2. Convert the original columns to Int64 using a safer 'round-trip' method
# This avoids the "safe cast" error by handling the NaN gap explicitly
df_filled['Quantity'] = qty_clean.round(0).astype('Int64')
df_filled['Price Per Unit'] = price_clean.round(0).astype('Int64')

# 3. Clean 'Total Spent' (Convert strings to NaN)
df_filled['Total Spent'] = pd.to_numeric(df_filled['Total Spent'], errors='coerce')

# 4. Calculate the product
# Since both are now 'Int64', the result will be numeric/Int64
calculated_spent = df_filled['Quantity'] * df_filled['Price Per Unit']

# 5. Fill the gaps
df_filled['Total Spent'] = df_filled['Total Spent'].fillna(calculated_spent)


In [ ]:
df_filled

Transaction ID      Item  Quantity  Price Per Unit  Total Spent  \
0       TXN_1961373    Coffee         2               2          4.0   
1       TXN_4977031      Cake         4               3         12.0   
2       TXN_4271903    Cookie         4               1          4.0   
3       TXN_7034554     Salad         2               5         10.0   
4       TXN_3160411    Coffee         2               2          4.0   
...             ...       ...       ...             ...          ...   
9995    TXN_7672686    Coffee         2               2          4.0   
9996    TXN_9659401       NaN         3            <NA>          3.0   
9997    TXN_5255387    Coffee         4               2          8.0   
9998    TXN_7695629    Cookie         3               1          3.0   
9999    TXN_6170729  Sandwich         3               4         12.0   

      Payment Method  Location Transaction Date  
0        Credit Card  Takeaway       2023-09-08  
1               Cash  In-store       2023-05-16  
2        Credit Card  In-store       2023-07-19  
3            UNKNOWN   UNKNOWN       2023-04-27  
4     Digital Wallet  In-store       2023-06-11  
...              ...       ...              ...  
9995             NaN   UNKNOWN       2023-08-30  
9996  Digital Wallet       NaN       2023-06-02  
9997  Digital Wallet       NaN       2023-03-02  
9998  Digital Wallet       NaN       2023-12-02  
9999            Cash  In-store       2023-11-07  

[10000 rows x 8 columns]

In [ ]:
df_filled['Transaction Date'] = pd.to_datetime(df_filled['Transaction Date'], errors='coerce')

In [ ]:
df_filled.describe()

Quantity  Price Per Unit  Total Spent               Transaction Date
count    9521.0          9946.0       9977.0                           9540
mean   3.028463        3.007943     8.938408  2023-07-01 23:00:31.698113280
min         1.0             1.0          1.0            2023-01-01 00:00:00
25%         2.0             2.0          4.0            2023-04-01 00:00:00
50%         3.0             3.0          8.0            2023-07-02 00:00:00
75%         4.0             4.0         12.0            2023-10-02 00:00:00
max         5.0             5.0         25.0            2023-12-31 00:00:00
std    1.419007        1.221492     6.001686                            NaN

In [ ]:
df_filled.head(50)

Transaction ID      Item  Quantity  Price Per Unit  Total Spent  \
0     TXN_1961373    Coffee         2               2          4.0   
1     TXN_4977031      Cake         4               3         12.0   
2     TXN_4271903    Cookie         4               1          4.0   
3     TXN_7034554     Salad         2               5         10.0   
4     TXN_3160411    Coffee         2               2          4.0   
5     TXN_2602893  Smoothie         5               4         20.0   
6     TXN_4433211      Cake         3               3          9.0   
7     TXN_6699534  Sandwich         4               4         16.0   
8     TXN_4717867      Cake         5               3         15.0   
9     TXN_2064365  Sandwich         5               4         20.0   
10    TXN_2548360     Salad         5               5         25.0   
11    TXN_3051279  Sandwich         2               4          8.0   
12    TXN_7619095  Sandwich         2               4          8.0   
13    TXN_9437049    Cookie         5               1          5.0   
14    TXN_8915701       Tea         2               2          3.0   
15    TXN_2847255     Salad         3               5         15.0   
16    TXN_3765707  Sandwich         1               4          4.0   
17    TXN_6769710     Juice         2               3          6.0   
18    TXN_8876618      Cake         5               3         15.0   
19    TXN_3709394     Juice         4               3         12.0   
20    TXN_3522028  Smoothie      <NA>               4         20.0   
21    TXN_3567645  Smoothie         4               4         16.0   
22    TXN_5132361  Sandwich         3               4         12.0   
23    TXN_2616390  Sandwich         2               4          8.0   
24    TXN_9400181  Sandwich         5               4         20.0   
25    TXN_7958992  Smoothie         3               4         12.0   
26    TXN_5183041    Cookie         5               1          5.0   
27    TXN_5695074     Juice         4               3         12.0   
28    TXN_8467949  Smoothie         5               4         20.0   
29    TXN_7640952      Cake         4               3         12.0   
30    TXN_1736287    Coffee         5               2         10.0   
31    TXN_8927252    Cookie         2               1          2.0   
32    TXN_9677376  Smoothie         4               4         16.0   
33    TXN_7710508    Cookie         5               1          5.0   
34    TXN_8853997  Smoothie         2               4          8.0   
35    TXN_9130559  Sandwich         1               4          4.0   
36    TXN_6855453      Cake         4               3         12.0   
37    TXN_1080432     Salad         2               5         10.0   
38    TXN_2655815  Smoothie         4               4         16.0   
39    TXN_6688524    Coffee         4               2          8.0   
40    TXN_2083138  Smoothie         3               4         12.0   
41    TXN_2427584  Sandwich         4               4         16.0   
42    TXN_6650263       Tea         2               2          4.0   
43    TXN_9620080     Juice         4               3         12.0   
44    TXN_1491578    Cookie         2               1          2.0   
45    TXN_5455792     Salad         3               5         15.0   
46    TXN_8078640     Juice         4               3         12.0   
47    TXN_9499313     Juice         5               3         15.0   
48    TXN_8201146     Juice         5               3         15.0   
49    TXN_8230936      Cake         3               3          9.0   

    Payment Method  Location Transaction Date  
0      Credit Card  Takeaway       2023-09-08  
1             Cash  In-store       2023-05-16  
2      Credit Card  In-store       2023-07-19  
3          UNKNOWN   UNKNOWN       2023-04-27  
4   Digital Wallet  In-store       2023-06-11  
5      Credit Card       NaN       2023-03-31  
6            ERROR  Takeaway       2023-10-06  
7             Cash   UNKNOWN       2023-10-28  


In [ ]:
unique_quantities = df_filled['Quantity'].unique()
print(unique_quantities)

<IntegerArray>
[2, 4, 5, 3, 1, <NA>]
Length: 6, dtype: Int64


In [ ]:
df_filled.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    10000 non-null  object        
 1   Item              9977 non-null   object        
 2   Quantity          9521 non-null   Int64         
 3   Price Per Unit    9946 non-null   Int64         
 4   Total Spent       9977 non-null   Float64       
 5   Payment Method    7421 non-null   object        
 6   Location          6735 non-null   object        
 7   Transaction Date  9540 non-null   datetime64[ns]
dtypes: Float64(1), Int64(2), datetime64[ns](1), object(4)
memory usage: 654.4+ KB


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Transaction ID    10000 non-null  object
 1   Item              9667 non-null   object
 2   Quantity          9862 non-null   object
 3   Price Per Unit    9821 non-null   object
 4   Total Spent       9827 non-null   object
 5   Payment Method    7421 non-null   object
 6   Location          6735 non-null   object
 7   Transaction Date  9841 non-null   object
dtypes: object(8)
memory usage: 625.1+ KB


In [ ]:
uniques_L =  df_filled['Payment Method'].unique()
print(uniques_L)

['Credit Card' 'Cash' 'UNKNOWN' 'Digital Wallet' 'ERROR' nan]


- #### We have recovered some of the data in following columns **{Item, Price Per Unit, Total Spent}** by recognizing the pattern between them.
- #### However there is no direct indirect pattern or relation in columns **{Payment Method, Location}**

## Solving the Problem
Method Two: Proportional Random Imputation
### Logic Overview
This method treats the existing valid data as a statistical representative of the whole.

- Step A: Identify the "Noise" (NaN, UNKNOWN, ERROR) and isolate the "Valid" data.

- Step B: Calculate the Relative Frequency (Ratio) of the valid categories.

- Step C: Generate a random list of choices based on those ratios to fill the noise, ensuring the final dataset maintains the same statistical balance.

In [ ]:
import pandas as pd
import numpy as np

def fill_location_and_payment(df):
    """
    Fill UNKNOWN, NaN, and ERROR values in Location and Payment Method
    based on the ratio of valid values, randomly distributed.
    """

    df_filled = df.copy()

    # ========================================================================
    # STEP 1: Fill Location
    # ========================================================================

    # Get valid locations (exclude UNKNOWN, NaN, ERROR)
    valid_locations = df_filled['Location'][
        ~df_filled['Location'].isin(['UNKNOWN', 'ERROR']) &
        df_filled['Location'].notna()
    ]

    # Count occurrences of each valid location
    location_counts = valid_locations.value_counts()

    # Calculate ratio/probabilities
    location_probabilities = location_counts / location_counts.sum()

    print("Location Distribution:")
    print(location_counts)
    print("\nLocation Probabilities:")
    print(location_probabilities)
    print()

    # Find rows with invalid location
    invalid_location_mask = (
        df_filled['Location'].isin(['UNKNOWN', 'ERROR']) |
        df_filled['Location'].isna()
    )

    num_invalid_locations = invalid_location_mask.sum()
    print(f"Total invalid locations to fill: {num_invalid_locations}")

    # Generate random locations based on ratio
    random_locations = np.random.choice(
        location_probabilities.index,
        size=num_invalid_locations,
        p=location_probabilities.values
    )

    # Fill invalid locations
    df_filled.loc[invalid_location_mask, 'Location'] = random_locations

    print(f"✓ Filled {num_invalid_locations} locations")
    print()

    # ========================================================================
    # STEP 2: Fill Payment Method
    # ========================================================================

    # Get valid payment methods (exclude UNKNOWN, NaN, ERROR)
    valid_payments = df_filled['Payment Method'][
        ~df_filled['Payment Method'].isin(['UNKNOWN', 'ERROR']) &
        df_filled['Payment Method'].notna()
    ]

    # Count occurrences of each valid payment method
    payment_counts = valid_payments.value_counts()

    # Calculate ratio/probabilities
    payment_probabilities = payment_counts / payment_counts.sum()

    print("Payment Method Distribution:")
    print(payment_counts)
    print("\nPayment Method Probabilities:")
    print(payment_probabilities)
    print()

    # Find rows with invalid payment method
    invalid_payment_mask = (
        df_filled['Payment Method'].isin(['UNKNOWN', 'ERROR']) |
        df_filled['Payment Method'].isna()
    )

    num_invalid_payments = invalid_payment_mask.sum()
    print(f"Total invalid payment methods to fill: {num_invalid_payments}")

    # Generate random payment methods based on ratio
    random_payments = np.random.choice(
        payment_probabilities.index,
        size=num_invalid_payments,
        p=payment_probabilities.values
    )

    # Fill invalid payment methods
    df_filled.loc[invalid_payment_mask, 'Payment Method'] = random_payments

    print(f"✓ Filled {num_invalid_payments} payment methods")
    print()

    return df_filled


# Usage:
# df_filled = fill_location_and_payment(df)

In [ ]:
df_filled = fill_location_and_payment(df_filled)
df_filled.head(20)

Location Distribution:
Location
Takeaway    3022
In-store    3017
Name: count, dtype: int64

Location Probabilities:
Location
Takeaway    0.500414
In-store    0.499586
Name: count, dtype: float64

Total invalid locations to fill: 3961
✓ Filled 3961 locations

Payment Method Distribution:
Payment Method
Digital Wallet    2291
Credit Card       2273
Cash              2258
Name: count, dtype: int64

Payment Method Probabilities:
Payment Method
Digital Wallet    0.335825
Credit Card       0.333187
Cash              0.330988
Name: count, dtype: float64

Total invalid payment methods to fill: 3178
✓ Filled 3178 payment methods



Transaction ID      Item  Quantity  Price Per Unit  Total Spent  \
0     TXN_1961373    Coffee         2               2          4.0   
1     TXN_4977031      Cake         4               3         12.0   
2     TXN_4271903    Cookie         4               1          4.0   
3     TXN_7034554     Salad         2               5         10.0   
4     TXN_3160411    Coffee         2               2          4.0   
5     TXN_2602893  Smoothie         5               4         20.0   
6     TXN_4433211      Cake         3               3          9.0   
7     TXN_6699534  Sandwich         4               4         16.0   
8     TXN_4717867      Cake         5               3         15.0   
9     TXN_2064365  Sandwich         5               4         20.0   
10    TXN_2548360     Salad         5               5         25.0   
11    TXN_3051279  Sandwich         2               4          8.0   
12    TXN_7619095  Sandwich         2               4          8.0   
13    TXN_9437049    Cookie         5               1          5.0   
14    TXN_8915701       Tea         2               2          3.0   
15    TXN_2847255     Salad         3               5         15.0   
16    TXN_3765707  Sandwich         1               4          4.0   
17    TXN_6769710     Juice         2               3          6.0   
18    TXN_8876618      Cake         5               3         15.0   
19    TXN_3709394     Juice         4               3         12.0   

    Payment Method  Location Transaction Date  
0      Credit Card  Takeaway       2023-09-08  
1             Cash  In-store       2023-05-16  
2      Credit Card  In-store       2023-07-19  
3      Credit Card  In-store       2023-04-27  
4   Digital Wallet  In-store       2023-06-11  
5      Credit Card  In-store       2023-03-31  
6   Digital Wallet  Takeaway       2023-10-06  
7             Cash  Takeaway       2023-10-28  
8             Cash  Takeaway       2023-07-28  
9      Credit Card  In-store       2023-12-31  
10            Cash  Takeaway       2023-11-07  
11     Credit Card  Takeaway              NaT  
12            Cash  In-store       2023-05-03  
13            Cash  Takeaway       2023-06-01  
14     Credit Card  In-store       2023-03-21  
15     Credit Card  In-store       2023-11-15  
16            Cash  Takeaway       2023-06-10  
17            Cash  In-store       2023-02-24  
18            Cash  In-store       2023-03-25  
19            Cash  Takeaway       2023-01-15

In [ ]:
df_filled.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    10000 non-null  object        
 1   Item              9977 non-null   object        
 2   Quantity          9521 non-null   Int64         
 3   Price Per Unit    9946 non-null   Int64         
 4   Total Spent       9977 non-null   Float64       
 5   Payment Method    10000 non-null  object        
 6   Location          10000 non-null  object        
 7   Transaction Date  9540 non-null   datetime64[ns]
dtypes: Float64(1), Int64(2), datetime64[ns](1), object(4)
memory usage: 654.4+ KB


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Transaction ID    10000 non-null  object
 1   Item              9667 non-null   object
 2   Quantity          9862 non-null   object
 3   Price Per Unit    9821 non-null   object
 4   Total Spent       9827 non-null   object
 5   Payment Method    7421 non-null   object
 6   Location          6735 non-null   object
 7   Transaction Date  9841 non-null   object
dtypes: object(8)
memory usage: 625.1+ KB


In [ ]:
df.head(20)

Transaction ID      Item Quantity Price Per Unit Total Spent  \
0     TXN_1961373    Coffee        2              2           4   
1     TXN_4977031      Cake        4              3          12   
2     TXN_4271903    Cookie        4              1       ERROR   
3     TXN_7034554     Salad        2              5          10   
4     TXN_3160411    Coffee        2              2           4   
5     TXN_2602893  Smoothie        5              4          20   
6     TXN_4433211   UNKNOWN        3              3           9   
7     TXN_6699534  Sandwich        4              4          16   
8     TXN_4717867       NaN        5              3          15   
9     TXN_2064365  Sandwich        5              4          20   
10    TXN_2548360     Salad        5              5          25   
11    TXN_3051279  Sandwich        2              4           8   
12    TXN_7619095  Sandwich        2              4           8   
13    TXN_9437049    Cookie        5              1           5   
14    TXN_8915701     ERROR        2            1.5           3   
15    TXN_2847255     Salad        3              5          15   
16    TXN_3765707  Sandwich        1              4           4   
17    TXN_6769710     Juice        2              3           6   
18    TXN_8876618      Cake        5              3          15   
19    TXN_3709394     Juice        4              3          12   

    Payment Method  Location Transaction Date  
0      Credit Card  Takeaway       2023-09-08  
1             Cash  In-store       2023-05-16  
2      Credit Card  In-store       2023-07-19  
3          UNKNOWN   UNKNOWN       2023-04-27  
4   Digital Wallet  In-store       2023-06-11  
5      Credit Card       NaN       2023-03-31  
6            ERROR  Takeaway       2023-10-06  
7             Cash   UNKNOWN       2023-10-28  
8              NaN  Takeaway       2023-07-28  
9              NaN  In-store       2023-12-31  
10            Cash  Takeaway       2023-11-07  
11     Credit Card  Takeaway            ERROR  
12            Cash  In-store       2023-05-03  
13             NaN  Takeaway       2023-06-01  
14             NaN  In-store       2023-03-21  
15     Credit Card  In-store       2023-11-15  
16             NaN       NaN       2023-06-10  
17            Cash  In-store       2023-02-24  
18            Cash     ERROR       2023-03-25  
19            Cash  Takeaway       2023-01-15

In [ ]:

def compare_completeness(df, df_filled, name1="Dataset A", name2="Dataset B"):
    # List of values we consider as "Null"
    null_types = [np.nan, None, 'UNKNOWN', 'ERROR', 'nan', 'null']

    def get_stats(ddf):
        # Create a mask where True = valid data, False = any type of null
        valid_mask = ~ddf.isin(null_types) & df.notna()

        # Calculate count of valid rows and the percentage of completeness
        stats = {
            'Valid Count': valid_mask.sum(),
            'Completeness %': (valid_mask.sum() / len(df) * 100).round(2)
        }
        return pd.DataFrame(stats)

    # Get stats for both dataframes
    stats1 = get_stats(df)
    stats2 = get_stats(df_filled)

    # Merge side-by-side
    comparison = pd.concat([stats1, stats2], axis=1, keys=[name1, name2])

    return comparison

# --- Execution ---
# Assuming df_original and df_filled are your two datasets
audit_report = compare_completeness(df, df_filled, "Before Cleaning", "After Cleaning")

print("--- Data Completeness Comparison Report ---")
print(audit_report)

--- Data Completeness Comparison Report ---
                 Before Cleaning                After Cleaning               
                     Valid Count Completeness %    Valid Count Completeness %
Transaction ID             10000         100.00          10000          100.0
Item                        9031          90.31           9636          96.36
Quantity                    9521          95.21           9862          98.62
Price Per Unit              9467          94.67           9821          98.21
Total Spent                 9498          94.98           9827          98.27
Payment Method              6822          68.22           7421          74.21
Location                    6039          60.39           6735          67.35
Transaction Date            9540          95.40           9841          98.41
